File: src/csharp/Intel/Iced.UnitTests/Intel/FormatterTests/OptionsTestsUtils.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if GAS || INTEL || MASM || NASM || FAST_FMT
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Iced.UnitTests.Intel.FormatterTests {
	static class OptionsTestsUtils {
		public static IEnumerable<object[]> GetFormatData_Common(string formatterDir, string formattedStringsFile) {
			var (tests, ignored) = FormatterOptionsTests.CommonTests;
			return GetFormatData(formatterDir, formattedStringsFile, tests, ignored);
		}

		public static IEnumerable<object[]> GetFormatData_All(string formatterDir, string formattedStringsFile) {
			var (tests, ignored) = FormatterOptionsTests.AllTests;
			return GetFormatData(formatterDir, formattedStringsFile, tests, ignored);
		}

		public static IEnumerable<object[]> GetFormatData(string formatterDir, string formattedStringsFile, string optionsFile) {
			var testsFilename = FileUtils.GetFormatterFilename(Path.Combine(formatterDir, optionsFile));
			var ignored = new HashSet<int>();
			var tests = OptionsTestsReader.ReadFile(testsFilename, ignored).ToArray();
			return GetFormatData(formatterDir, formattedStringsFile, tests, ignored);
		}

		static IEnumerable<object[]> GetFormatData(string formatterDir, string formattedStringsFile, OptionsTestCase[] tests, HashSet<int> ignored) {
			var formattedStrings = FileUtils.ReadRawStrings(Path.Combine(formatterDir, formattedStringsFile)).ToArray();
			formattedStrings = Utils.Filter(formattedStrings, ignored);
			if (tests.Length != formattedStrings.Length)
				throw new ArgumentException($"(tests.Length) {tests.Length} != (formattedStrings.Length) {formattedStrings.Length} . tests[0].HexBytes = {(tests.Length == 0 ? "<EMPTY>" : tests[0].HexBytes)} & formattedStrings[0] = {(formattedStrings.Length == 0 ? "<EMPTY>" : formattedStrings[0])}");
			var res = new object[tests.Length][];
			for (int i = 0; i < tests.Length; i++)
				res[i] = new object[3] { i, tests[i], formattedStrings[i] };
			return res;
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/fmt/OptionsTestsUtils.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt;

import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashSet;

import org.junit.jupiter.params.provider.Arguments;

final class OptionsTestsUtils {
	static Iterable<Arguments> getFormatData_Common(String formatterDir, String formattedStringsFile) {
		FormatterOptionsTests.Info info = FormatterOptionsTests.commonTests;
		return getFormatData(formatterDir, formattedStringsFile, info.tests, info.ignored);
	}

	static Iterable<Arguments> getFormatData_All(String formatterDir, String formattedStringsFile) {
		FormatterOptionsTests.Info info = FormatterOptionsTests.allTests;
		return getFormatData(formatterDir, formattedStringsFile, info.tests, info.ignored);
	}

	static Iterable<Arguments> getFormatData(String formatterDir, String formattedStringsFile, String optionsFile) {
		String testsFilename = FmtFileUtils.getFormatterFilename(Paths.get(formatterDir, optionsFile).toString());
		HashSet<Integer> ignored = new HashSet<Integer>();
		OptionsTestCase[] tests = OptionsTestsReader.readFile(testsFilename, ignored).toArray(new OptionsTestCase[0]);
		return getFormatData(formatterDir, formattedStringsFile, tests, ignored);
	}

	private static Iterable<Arguments> getFormatData(String formatterDir, String formattedStringsFile, OptionsTestCase[] tests, HashSet<Integer> ignored) {
		String[] formattedStrings = FmtFileUtils.readRawStrings(Paths.get(formatterDir, formattedStringsFile).toString()).toArray(new String[0]);
		formattedStrings = Utils.filter(formattedStrings, ignored);
		if (tests.length != formattedStrings.length)
			throw new UnsupportedOperationException();
		ArrayList<Arguments> res = new ArrayList<Arguments>(tests.length);
		for (int i = 0; i < tests.length; i++)
			res.add(Arguments.of(i, tests[i], formattedStrings[i]));
		return res;
	}
}
